File: src/tags.ts
import { existsSync, readFileSync, writeFileSync } from 'fs';
import path from 'path';

interface Tag {
    name: string;
    songs: string[];
}

const tagsFileName = 'tags.json';

export function getTags(songsPath: string) {
    const filePath = path.join(songsPath, tagsFileName);

    if (!existsSync(filePath)) {
        return [];
    }

    return JSON.parse(readFileSync(filePath, 'utf-8') || '[]') as Tag[];
}

export function changeSongsInTag(
    songsPath: string,
    tagName: string,
    songs: string[],
    append: boolean
) {
    const tags = getTags(songsPath);
    const filePath = path.join(songsPath, tagsFileName);

    let tag = tags.find((t) => t.name === tagName);

    if (!tag) {
        tags.push({
            name: tagName,
            songs,
        });
    } else if (!append) {
        tag.songs = songs;
    } else {
        for (const song of songs) {
            if (!tag.songs.includes(song)) {
                tag.songs.push(song);
            }
        }
    }

    writeFileSync(filePath, JSON.stringify(tags), {});
}

File: commands/play/tags.go
package play

import (
	"encoding/json"
	"errors"
	"io/fs"
	"os"
	"path/filepath"
)

type Tag struct {
	Name  string   `json:"name"`
	Songs []string `json:"songs"`
}

func getTagPath(musicPath string) string {
	return filepath.Join(musicPath, "tags.json")
}

func getStoredTags(musicPath string) ([]Tag, error) {
	var savedTags []Tag

	content, err := os.ReadFile(getTagPath(musicPath))

	if err == nil {
		var payload []Tag
		err = json.Unmarshal(content, &payload)

		if err != nil {
			return nil, err
		}

		savedTags = payload
	} else if errors.Is(err, fs.ErrNotExist) {
		return savedTags, nil
	} else {
		return nil, err
	}

	return savedTags, nil
}

func changeSongsInTag(musicPath string, tagName string, songs []string, shouldAppend bool) error {
	var tag Tag

	tags, err := getStoredTags(musicPath)

	if err != nil {
		return err
	}

	for _, t := range tags {
		if t.Name == tagName {
			tag = t
			break
		}
	}

	if tag.Name == "" {
		tag = Tag{Name: tagName, Songs: make([]string, 0, len(songs))}
		tags = append(tags, tag)
	} else if !shouldAppend {
		tag.Songs = make([]string, 0, len(songs))
	}

	for _, song := range songs {
		bareName := getBareSongName(song, musicPath)

		if includes(tag.Songs, bareName) {
			tag.Songs = append(tag.Songs, bareName)
		}
	}

	tagsString, err := json.Marshal(tags)

	if err != nil {
		return err
	}

	err = os.WriteFile(getTagPath(musicPath), tagsString, 0666)
	return err
}
